### PR TITLE
NETOBSERV-909 fix servicemonitor & prom rule reconcile

### PR DESCRIPTION
**Pull request by jotak**
bug tld;dr: 
We need to be careful about not tying lifecycle of different objects together, as this would be assuming a certain state, which creates "holes" in the reconcile logic - we must avoid as much as possible any assumption about the current state

### CODE REVIEW

**Review by KalmanMeth — APPROVED** [review]

**Comment by memodi** [comment]
@jotak - I am seeing an error still this time with console-plugin reconcilation, if I tried procedure as described in ticket:
```
$ oc get flowcollector -w
NAME      AGENT   SAMPLING (EBPF)   DEPLOYMENT MODEL   STATUS
cluster   EBPF    1                 DIRECT             ReconcileConsolePluginFailed

2023-03-09T17:15:18.262Z	ERROR	controller.flowcollector	Failed to create new *v1.ClusterRoleBinding	{"reconciler group": "flows.netobserv.io", "reconciler kind": "FlowCollector", "name": "cluster", "namespace": "", "Namespace": "", "Name": "netobserv-plugin", "error": "clusterroles.rbac.authorization.k8s.io \"netobserv-plugin\" not found"}
github.com/netobserv/network-observability-operator/controllers/reconcilers.(*ClientHelper).ReconcileClusterRoleBinding
	/opt/app-root/controllers/reconcilers/client_helper.go:92
github.com/netobserv/network-observability-operator/controllers/consoleplugin.(*CPReconciler).reconcilePermissions
	/opt/app-root/controllers/consoleplugin/consoleplugin_reconciler.go:151
github.com/netobserv/network-observability-operator/controllers/consoleplugin.(*CPReconciler).Reconcile
	/opt/app-root/controllers/consoleplugin/consoleplugin_reconciler.go:97
github.com/netobserv/network-observability-operator/controllers.(*FlowCollectorReconciler).Reconcile
	/opt/app-root/controllers/flowcollector_controller.go:156
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/opt/app-root/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/opt/app-root/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/opt/app-root/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/opt/app-root/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227
2023-03-09T17:15:18.262Z	ERROR	controller.flowcollector	Failed to reconcile Console plugin: clusterroles.rbac.authorization.k8s.io "netobserv-plugin" not found	{"reconciler group": "flows.netobserv.io", "reconciler kind": "FlowCollector", "name": "cluster", "namespace": "", "error": "clusterroles.rbac.authorization.k8s.io \"netobserv-plugin\" not found"}
github.com/netobserv/network-observability-operator/controllers.(*FlowCollectorReconciler).Reconcile
	/opt/app-root/controllers/flowcollector_controller.go:158
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/opt/app-root/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/opt/app-root/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/opt/app-root/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/opt/app-root/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227
2023-03-09T17:15:18.293Z	ERROR	controller.flowcollector	Reconciler error	{"reconciler group": "flows.netobserv.io", "reconciler kind": "FlowCollector", "name": "cluster", "namespace": "", "error": "clusterroles.rbac.authorization.k8s.io \"netobserv-plugin\" not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/opt/app-root/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/opt/app-root/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227
```

**Comment by jotak** [comment]
@OlivierCazade can you take a second look please? @memodi pointed out an issue that turned out needing more changes, cf my last commit e0e9d0dee309d077ebc05c7bb93bcca58cb4645f

**Comment by jotak** [comment]
@memodi I retried and cannot reproduce the initial issue with this fix applied. Maybe we need to check if our reproduction scenario differ. What I do is:
1. Install downstream operator 1.1
2. Create a default flowcollector (without changing any value - also deploying 0-click loki)
3. Ensure everything works as expected
4. Delete the operator, keeping the FlowCollector installed
5. Deploy this PR operator using `IMG=... make deploy`
6. Check logs & readiness with `oc  get flowcollector`

**Comment by jotak** [comment]
Just retried using loki-operator instead of zero-click, to make sure it isn't related to that : still no issue

**Comment by memodi** [comment]
/label qe-approved

thanks @jotak - confirming both the issues  - Reconciliation during upgrade and NOO image with pre-merge catalog image - are fixed with most recent changes.

File: .github/workflows/pull_request.yml [hunk 1]
```diff
@@ -29,7 +29,7 @@ jobs:
       with:
         files: ./cover.out
         flags: unittests
-        fail_ci_if_error: true        
+        fail_ci_if_error: false
```
Comment (jpinsonneau) [inline .github/workflows/pull_request.yml:32]: Is that the solution to flaky test coverage ? Will it skip if it fails to upload for example ?

**Review by jpinsonneau — APPROVED** [review]
LGTM in terms of code (not tested)

File: .github/workflows/pull_request.yml (hunk 1 above)
Comment (jotak) [reply .github/workflows/pull_request.yml:32]: yeah well, IMO coverage upload is not something critical enough to make whole CI fail, given how unstable it is

**Pull request merged**